riscv_isa.cpp

```diff
@@ -1,489 +1,810 @@
 void ac_behavior(SUB) {
-  dbg_printf("SUB r%d, r%d, r%d", rd, rs1, rs2);
+  dbg_printf("SUB r%d, r%d, r%d\n", rd, rs1, rs2);
+  dbg_printf("RB[rs1] = %d\n", RB[rs1]);
+  dbg_printf("RB[rs2] = %d\n", RB[rs2]);
   RB[rd] = RB[rs1] - RB[rs2];
-  dbg_printf("Result = %#x\n", RB[rd]);
-  // RISCV doesn't check for overflow
+  dbg_printf("Result = %d\n\n", RB[rd]);
 }
 
+
+// Instruction SLL behavior method.
 void ac_behavior(SLL) {
-  dbg_printf("SLL r%d, r%d, r%d", rd, rs1, rs2);
-  RB[rd] = RB[rs1] << RB[rs2];
-  dbg_printf("Result = %#x\n", RB[rd]);
+  dbg_printf("SLL r%d, r%d, r%d\n", rd, rs1, rs2);
+  short int shamt = RB[rs2];
+  shamt = shamt << 11;
+  shamt = shamt >> 11;
+  RB[rd] = RB[rs1] << shamt;
+  dbg_printf("RB[rs1] = %d\n", RB[rs1]);
+  dbg_printf("RB[rs2] = %d\n", RB[rs2]);
+  dbg_printf("Result = %d\n\n", RB[rd]);
 }
 
+
+// Instruction SLT behavior method.
 void ac_behavior(SLT) {
-  dbg_printf("SLT r%d, r%d, r%d", rd, rs1, rs2);
-  if (RB[rs1] < RB[rs2])
+  dbg_printf("SLT r%d, r%d, r%d\n", rd, rs1, rs2);
+  if ((ac_Sword)RB[rs1] < (ac_Sword)RB[rs2])
     RB[rd] = 1;
   else
     RB[rd] = 0;
-  dbg_printf("Result = %#x\n", RB[rd]);
+  dbg_printf("RB[rs1] = %d\n", RB[rs1]);
+  dbg_printf("RB[rs2] = %d\n", RB[rs2]);
+  dbg_printf("Result = %d\n\n", RB[rd]);
 }
 
+
+// Instruction SLTU behavior method.
 void ac_behavior(SLTU) {
-  dbg_printf("SLTU r%d, r%d, r%d", rd, rs1, rs2);
+  dbg_printf("SLTU r%d, r%d, r%d\n", rd, rs1, rs2);
   if ((ac_Uword)RB[rs1] < (ac_Uword)RB[rs2])
     RB[rd] = 1;
   else
     RB[rd] = 0;
-  dbg_printf("Result = %#x\n", RB[rd]);
+  dbg_printf("RB[rs1] = %#x\n", RB[rs1]);
+  dbg_printf("RB[rs2] = %#x\n", RB[rs2]);
+  dbg_printf("Result = %#x\n\n", RB[rd]);
 }
 
+
+// Instruction XOR behavior method.
 void ac_behavior(XOR) {
-  dbg_printf("XOR r%d, r%d, r%d", rd, rs1, rs2);
+  dbg_printf("XOR r%d, r%d, r%d\n", rd, rs1, rs2);
   RB[rd] = RB[rs1] ^ RB[rs2];
-  dbg_printf("Result = %#x\n", RB[rd]);
+  dbg_printf("RB[rs1] = %#x\n", RB[rs1]);
+  dbg_printf("RB[rs2] = %#x\n", RB[rs2]);
+  dbg_printf("Result = %#x\n\n", RB[rd]);
 }
 
+
+// Instruction SRL behavior method.
 void ac_behavior(SRL) {
-  dbg_printf("SRL r%d, r%d, r%d", rd, rs1, rs2);
-  RB[rd] = RB[rs1] >> RB[rs2];
-  dbg_printf("Result = %#x\n", RB[rd]);
+  dbg_printf("SRL r%d, r%d, r%d\n", rd, rs1, rs2);
+  short int shamt = RB[rs2];
+  shamt = shamt << 11;
+  shamt = shamt >> 11;
+  RB[rd] = RB[rs1] >> shamt;
+  dbg_printf("RB[rs1] = %d\n", RB[rs1]);
+  dbg_printf("RB[rs2] = %d\n", RB[rs2]);
+  dbg_printf("Result = %d\n\n", RB[rd]);
 }
 
+
+// Instruction SRA behavior method.
 void ac_behavior(SRA) {
-  dbg_printf("SRA r%d, r%d, r%d", rd, rs1, rs2);
-  if ((RB[rs1] >> 4) == 1)
-    RB[rd] = (RB[rs1] >> RB[rs2]) | (0xFFFFFFFF << RB[rs2]);
-  else
-    RB[rd] = RB[rs1] >> RB[rs2];
-  dbg_printf("Result = %#x\n", RB[rd]);
+  dbg_printf("SRA r%d, r%d, r%d\n", rd, rs1, rs2);
+  short int shamt = RB[rs2];
+  shamt = shamt << 11;
+  shamt = shamt >> 11;
```

Comment:
I couldn't understand what you did here. Could you explain? Aren't you supposed to just use the lower 5 bits of RB[rs2]?